File: guests/haskell/commands.hs
{-# LANGUAGE DataKinds #-}
{-# LANGUAGE OverloadedStrings #-}

-- The commands scene, Haskell port — guests/rust/commands.rs,
-- tools/scenes/commands.steps.

import Data.IORef (modifyIORef', newIORef, readIORef)

import KayaApp

main :: IO ()
main = kayaMain $ \app -> do
  settingsRef <- newIORef (0 :: Int)

  buildTx app $ do
    status <- signalText "ready"
    details <- signalBool False
    sort <- signalDouble 0.0

    let onSettings = do
          modifyIORef' settingsRef (+ 1)
          n <- readIORef settingsRef
          submitTx app (writeSignal status ("settings " <> tshow n))

    window
      primary
      [ WTitle "commands",
        WMenus
          [ -- Reload sits beside Settings so this menu is not left empty
            -- once the platform moves the settings item elsewhere.
            menu
              "File"
              []
              [ item "Reload" [],
                item
                  "Settings…"
                  [ IShortcut "primary+comma",
                    IRole roleSettings,
                    IOnActivate onSettings
                  ]
              ],
            -- Option order IS the index vocabulary: Name = 0, Date = 1.
            menu
              "View"
              []
              [ toggle
                  "Details"
                  [ ICheckedBy details,
                    IShortcut "primary+backslash",
                    IOnToggle
                      ( \on ->
                          submitTx app $
                            writeSignal status
                              (if on then "details on" else "details off")
                      )
                  ],
                radioGroup
                  "Sort"
                  [ IValueBy sort,
                    IOnSelect
                      ( \index ->
                          submitTx app $
                            writeSignal status
                              (if index == 1 then "sorted date" else "sorted name")
                      )
                  ]
                  [ option "Name" [IShortcut "primary+1"],
                    option "Date" [IShortcut "primary+2"]
                  ]
              ]
          ]
      ]

    root <- column [] [labelBound status] -- label#0
    mount root

File: guests/java/dev/kaya/guests/Commands.java
package dev.kaya.guests;

import dev.kaya.KayaApp;

/**
 * The commands scene from the JVM — guests/rust/commands.rs,
 * tools/scenes/commands.steps.
 */
public final class Commands {
    // Java lambdas cannot assign captured locals.
    private static int settingsCount;

    public static void app() {
        KayaApp app = new KayaApp();
        settingsCount = 0;

        app.build(tx -> {
            KayaApp.Signal<String> status = tx.signal("ready");
            KayaApp.Signal<Boolean> details = tx.signal(false);
            KayaApp.Signal<Double> sort = tx.signal(0.0);

            KayaApp.WindowRef win = tx.window(0).title("commands");

            // Reload keeps this menu non-empty once macOS moves Settings out.
            KayaApp.MenuItem file = win.menu("File");
            file.item("Reload");
            file.item("Settings…").shortcut("primary+comma").role(KayaApp.ROLE_SETTINGS)
                    .onActivate(t -> {
                        // Fires twice on purpose: the chord and the declared path.
                        settingsCount++;
                        t.write(status, "settings " + settingsCount);
                    });

            KayaApp.MenuItem view = win.menu("View");
            view.toggle("Details").checked(details).shortcut("primary+backslash")
                    .onToggle((t, on) -> t.write(status, on ? "details on" : "details off"));

            // Option order IS the index vocabulary: Name = 0, Date = 1.
            KayaApp.MenuItem sortGroup = view.radioGroup("Sort");
            sortGroup.option("Name").shortcut("primary+1");
            sortGroup.option("Date").shortcut("primary+2");
            sortGroup.value(sort).onSelect((t, index) ->
                    t.write(status, index == 1 ? "sorted date" : "sorted name"));

            tx.mount(tx.column(() -> {
                tx.label(status); // label#0
            }));
        });

        app.dispatchLoop();
    }
}
